narwhal/types/src/primary.rs

```diff
@@ -376,18 +376,7 @@ impl Hash<{ crypto::DIGEST_LENGTH }> for Header {
 
     fn digest(&self) -> HeaderDigest {
         let mut hasher = crypto::DefaultHashFunction::new();
-        hasher.update(&self.author);
-        hasher.update(self.round.to_le_bytes());
-        hasher.update(self.epoch.to_le_bytes());
-        hasher.update(self.created_at.to_le_bytes());
-        for (x, (y, z)) in self.payload.iter() {
-            hasher.update(Digest::from(*x));
-            hasher.update(y.to_le_bytes());
-            hasher.update(z.to_le_bytes());
-        }
-        for x in self.parents.iter() {
-            hasher.update(Digest::from(*x))
-        }
+        hasher.update(bcs::to_bytes(&self).expect("Serialization should not fail"));
```

Comment:
since header already derives serialization, this should be sufficient?

Reply:
Hm interesting, what I realize from previous implementation is we don't serialize everything, ie the signature part.

Reply:
i just came across @aschran's commit - https://github.com/MystenLabs/sui/commit/97954cb78b2b3494234c1210785d5c9fa75359bb#diff-55bd5a5412329cc279a55b6c1a86c518048159dbf0b546f5b882413cee567965

do we want to use bcs to compute hash for header digest as well?

Reply:
@joyqvq I'm curious, what is the advantage of using bcs here, besides standardization?

After a discussion with @kchalkias and @andll, I thought a bit more and realized maybe we can remove the `signature` field from `Header`: each voter should just verify header.author is the same as the anemo peer that requested the header to be signed. Maybe we can propose this as a separate change.

Reply:
Re "why use bcs" - one benefit is that with this approach we are not prone to errors like "we added a field but forgot to update hash calculation function" (which can happen and has very dire consequences). Also I do think standardization is a good thing (though even with this change we seem to do things differently in sui and narwhal, but I guess one step at a time)

Reply:
agree on both 1) bcs for not needing to think about the new fields 2) removing sig from the header

@andll you mean the difference with sui is:
1) sui uses sha3_256 whereas narwhal uses blake
2) sui uses the signable trait where the struct name is prepended when calculating hash
```
pub fn sha3_hash<S: Signable<Sha3_256>>(signable: &S) -> [u8; 32] {
    let mut digest = Sha3_256::default();
    signable.write(&mut digest);
    let hash = digest.finalize();
    hash.into()
}
```
1) is for efficiency since the digest in narwhal is not client facing. 2) is the BCSSignable something we want to introduce to narwhal?

Reply:
Yeah, those two things :)
1. If the different hash is the performance choice then we should keep it, thanks for explaining :)
2. Yeah, I think eventually we could do this. But to be clear, I don't think this need to change in this PR(and its not something urgent by any means)